### PR TITLE
Add support for Customer Sales Tax and Cash Rounding

### PR DESCRIPTION
**Pull request by SteveIreland**
I'm closing my eyes and pulling the trigger on this.

It is a significant change so please test early and often.  Each night I will address any issues that anyone finds.

### CODE REVIEW

**Comment by SteveIreland** [comment]
The "Merge Mangler" still lives.  I'm not sure where this came from, but on the configuration tab the new default origin tax code was being passed into the tax decimals function before it was displayed.  I have a fix ready to go for this (and anything else I stumble across) as soon as the feature is merged.

File: application/libraries/Sale_lib.php [hunk 1]
```diff
@@ -315,21 +317,23 @@ public function get_amount_due()
 		$precision = $this->CI->config->item('currency_decimals');
 		$rounded_due = bccomp(round($amount_due, $precision, PHP_ROUND_HALF_EVEN), 0, $precision);
 		// take care of rounding error introduced by round tripping payment amount to the browser
-		return $rounded_due == 0 ? 0 : $amount_due;
+		$final_amount_due = $rounded_due == 0 ? 0 : $amount_due;
+		error_log('>>>Sale_lib.get_amount_due sales_total-'.$sales_total
```
Comment (daN4cat) [inline application/libraries/Sale_lib.php]: Is this still required?

File: application/controllers/Login.php [hunk 2]
```diff
@@ -71,7 +72,7 @@ public function login_check($username)
 	
 	private function _security_check($username, $password)
 	{
-		return preg_match('~\b(Copyright|(c)|�|All rights reserved|Developed|Crafted|Implemented|Made|Powered|Code|Design|unblockUI|blockUI|blockOverlay|hide|opacity)\b~i', file_get_contents(APPPATH . 'views/partial/footer.php'));
+		return preg_match('~\b(Copyright|(c)|�|All rights reserved|Developed|Crafted|Implemented|Made|Powered|Code|Design|unblockUI|blockUI|blockOverlay|hide|opacity)\b~i', file_get_contents(APPPATH . 'views/partial/footer.php'));
```
Comment (daN4cat) [inline application/controllers/Login.php]: please revert this change

File: application/libraries/Sale_lib.php [hunk 3]
```diff
@@ -757,6 +762,12 @@ public function copy_entire_sale($sale_id)
 			$this->add_payment($row->payment_type, $row->payment_amount);
 		}
 
+		// TODO This needs to be fleshed out
```
Comment (daN4cat) [inline application/libraries/Sale_lib.php]: what about this?

**Comment by daN4cat** [comment]
@SteveIreland there is a lot to review and I didn't have the time today. So I just did a quick check and a couple of comments.

Also I would need to check the old and new tables, and make sure the upgrade path is fine. But more in the next days unless @jekkos finds the time to review it before me.

**Comment by jekkos** [comment]
This change is huge. I'll have to take some time to review thoroughly. However a unit test (for tax_library?) that verifies some basic taxation scenarios would be welcomed here.  Think it is a good idea to move all tax related logic to it's own module (out of sale and receiving_lib)

**Comment by SteveIreland** [comment]
So "slamming it in" is out of the question?  😈 

Okay ... slow and easy is fine by me ...  just call me Mellow Yellow.

**Comment by daN4cat** [comment]
The point is that slamming it in and not having active testers will not help.
Also I would like to get my mind around this a bit to make sure we understand the design change as in the end we need to maintain if we want to keep the project alive.

**Comment by odiea** [comment]
How and where are these new tax codes to be setup?

**Comment by daN4cat** [comment]
@odiea check the permissions and enable the tax module. That's what I read from the code and even judging by the new icon addition.

**Comment by odiea** [comment]
That was the first place I checked and there is no new checkbox for taxes that I can see.

**Comment by daN4cat** [comment]
First test failed 😊 @SteveIreland upgrade process not working....

**Comment by odiea** [comment]
Enabled Customer Tax
Created 1 tax Under taxes
Set origin tax to 1
Tax does not show in sale without a customer selected.
Complete Sale and receipt is not shown it goes to a blank page. Disable Customer Tax. No sales tax charged but now the receipt shows.

**Comment by SteveIreland** [comment]
@odiea I heard rumors you were going to be gone for a couple of months.  :grin:  There is no way now that I can slip this past @jekkos and @daN4cat 

Seriously though, thanks for taking a look at it.

**Comment by SteveIreland** [comment]
My bad about database.sql.  I mind was focused on the version to version upgrade script and tables.sql to the point that  I neglected the database.sql.  In my defense (as inadequate as it is) I think I had in the back of my mind that the database.sql script was supposed to be generated from the tables.sql script.  Unfortunately if that's possible I haven't figured out how to do that yet.

**Comment by odiea** [comment]
Ya. I am in Honduras to reclaim a property we have. It is back to being  livable now so I am allowed to play with the boys again.

**Comment by SteveIreland** [comment]
I updated database.sql and attempted to fix a few other things.  

An odd situation with that login.php though ... I think character encoding is biting me on this.  I fixed the character so that it displays the same (before and after) ... at least on the page I see ... but the diff page still thinks there is a difference.  I never touched that line of code before this ... so it seems to be a change that was contributed by different development platforms or IDE's.

**Comment by daN4cat** [comment]
It's OK, don't worry about the login.php.
With regards to generate database.sql out of tables.sql, as I said you need to read the wiki and install npm grunt and bower.

**Comment by SteveIreland** [comment]
Just a note, I will be updating the wiki page https://github.com/jekkos/opensourcepos/wiki/Sales-Tax-by-Customer over the coming week.

**Comment by SteveIreland** [comment]
I found an issue with the Takings panel in this pull request.  Working on it today.

**Comment by SteveIreland** [comment]
Okay, the takings issue is fixed.  I'll update the pull request today.

**Comment by odiea** [comment]
No receipt reprint in Takings.
It appears Issue #1134 is fixed with this new tax.

**Comment by SteveIreland** [comment]
The issues I know about in takings have been incorporated in the pull request now.

**Comment by odiea** [comment]
Receipts in Takings are working now.
The first report is the Detailed Sales

The second is self explained
2017-03-25 02_39_02-open source point of sale _ powered by ospos 3 1 0

File: application/controllers/Customers.php [hunk 4]
```diff
@@ -79,6 +81,15 @@ public function view($customer_id = -1)
 		$data['packages'] = $packages;
 		$data['selected_package'] = $info->package_id;
 
+		if ($customer_sales_tax_support == '1')
```
Comment (daN4cat) [inline application/controllers/Customers.php]: Probably all this if statement can be simplified to $data['customer_sales_tax_enabled'] = ! empty($customer_sales_tax_support);

See http://php.net/manual/en/function.empty.php

File: application/controllers/Items.php [hunk 5]
```diff
@@ -209,6 +210,25 @@ public function view($item_id = -1)
 		$data['suppliers'] = $suppliers;
 		$data['selected_supplier'] = $item_info->supplier_id;
 
+		$customer_sales_tax_support = $this->config->item('customer_sales_tax_support');
+		if($customer_sales_tax_support == '1')
```
Comment (daN4cat) [inline application/controllers/Items.php]: Same story here, !empty would do better here and cover more cases of '1' or TRUE.

File: application/controllers/Sales.php [hunk 6]
```diff
@@ -484,10 +486,12 @@ public function complete()
 			else
 			{
 				$data['invoice_number'] = $invoice_number;
-				$data['quote_number'] = $quote_number;
+				$data['sale_status'] = '0'; // Complete
 
 				// Save the data to the sales table
-				$data['sale_id_num'] = $this->Sale->save($data['cart'], $customer_id, $employee_id, $data['comments'], $invoice_number, $data['payments'], $data['dinner_table']);
+				$data['sale_id_num'] = $this->Sale->save($data['sale_status'], $data['cart'], $customer_id, $employee_id, $data['comments'], $invoice_number, $data["quote_number"], $data['payments'], $data['dinner_table'], $data['taxes']);
+				// TODO Replace the first with the later
```
Comment (daN4cat) [inline application/controllers/Sales.php]: What's the plan with the "TODO"? I can see another TODO above related to invoices and it's left there....

File: application/controllers/Sales.php [hunk 7]
```diff
@@ -509,16 +513,16 @@ public function complete()
 		}
 		elseif($this->sale_lib->is_quote_mode())
 		{
+			$invoice_number = null;
```
Comment (daN4cat) [inline application/controllers/Sales.php]: NULL not null, we use the version in capital letter and I made that change across the code some months ago to have consistency in the coding standard.

File: application/controllers/Sales.php [hunk 8]
```diff
@@ -509,16 +513,16 @@ public function complete()
...
 			$quote_number = $this->sale_lib->get_quote_number();
+
 			if($quote_number == null)
```
Comment (daN4cat) [inline application/controllers/Sales.php]: NULL not null, we use the version in capital letter and I made that change across the code some months ago to have consistency in the coding standard.

File: application/controllers/Sales.php [hunk 9]
```diff
@@ -509,16 +513,16 @@ public function complete()
...
 			{
 				// generate quote number
 				$quote_format = $this->config->item('sales_quote_format');
 				$quote_number = $this->token_lib->render($quote_format);
 			}
 
-			$invoice_number = null;
-
 			// TODO If duplicate quote then determine the number of employees and repeat until until success or tried the number of employees (if QSEQ was used).
```
Comment (daN4cat) [inline application/controllers/Sales.php]: another TODO....

File: application/controllers/Sales.php [hunk 10]
```diff
@@ -529,13 +533,19 @@ public function complete()
 			{
 				$data['invoice_number'] = $invoice_number;
 				$data['quote_number'] = $quote_number;
+				$data['sale_status'] = '1'; // Suspended
```
Comment (daN4cat) [inline application/controllers/Sales.php]: Maybe instead on numbers as enum it would be better to use the strings: suspended, complete, etc.

File: application/controllers/Sales.php [hunk 11]
```diff
@@ -1107,17 +1122,34 @@ public function suspend_quote($quote_number)
...
+	 * are now stored in the sales tables with a sale_status value of suspended.
+	 */
 	public function unsuspend()
 	{
 		$sale_id = $this->input->post('suspended_sale_id');
 		$this->sale_lib->clear_all();
-		$this->sale_lib->copy_entire_suspended_sale($sale_id);
-		$this->Sale_suspended->delete($sale_id);
+
+		if($sale_id > 0) {
```
Comment (daN4cat) [inline application/controllers/Sales.php]: { in the next line...

File: application/libraries/Tax_lib.php [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+	public function __construct()
+	{
+		$this->CI =& get_instance();
+		$this->CI->load->library('sale_lib');
+	}
+
+	public function get_tax_types()
+	{
+		return array(
+			'0' => $this->CI->lang->line('taxes_sales_tax'),
```
Comment (daN4cat) [inline application/libraries/Tax_lib.php]: it's probably better to have a meaningful enum using a string, it would get very confusing to use '0' and '1' across the code.

File: application/libraries/Sale_lib.php [hunk 13]
```diff
@@ -821,34 +847,59 @@ public function is_customer_taxable()
...
-		$taxes = array();
-
+		$register_mode = $this->CI->config->config['default_register_mode'];
+		$tax_decimals = $this->CI->config->config['tax_decimals'];
+		$customer_id = $this->get_customer();
+		$customer = $this->CI->Customer->get_info($customer_id);
+		$sales_taxes = array();
 		//Do not charge sales tax if we have a customer that is not taxable
-		if($this->is_customer_taxable())
+		if($customer->taxable or $customer_id == -1)
```
Comment (daN4cat) [inline application/libraries/Sale_lib.php]: we should use || instead or or

File: application/libraries/Sale_lib.php [hunk 14]
```diff
@@ -821,34 +847,59 @@ public function is_customer_taxable()
...
 
+				$tax_info = $this->CI->Item_taxes->get_info($item['item_id']);
+				$tax_group_sequence = 0;
 				foreach($tax_info as $tax)
 				{
-					$name = to_tax_decimals($tax['percent']) . '% ' . $tax['name'];
-					$tax_amount = $this->get_item_tax($item['quantity'], $item['price'], $item['discount'], $tax['percent']);
+					// This computes tax for each line item and adds it to the tax type total
+					$tax_group = (float)$tax['percent'] . '% ' . $tax['name'];
+					$tax_type = '0'; // vat
```
Comment (daN4cat) [inline application/libraries/Sale_lib.php]: here again please use an enumeration instead of '0'.

File: application/libraries/Tax_lib.php [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+	 */
+	public function get_sales_tax_for_amount($tax_basis, $tax_percentage, $rounding_code, $decimals)
+	{
+
+		$tax_fraction = bcdiv($tax_percentage, 100);
+
+		$tax_amount = bcmul($tax_basis, $tax_fraction);
+		$rounded_tax_amount = $tax_amount;
+
+		if($rounding_code == 0)
```
Comment (daN4cat) [inline application/libraries/Tax_lib.php]: Here another case for an enumeration

**Comment by daN4cat** [comment]
Maybe we should consider this for the enumerations: https://github.com/marc-mabe/php-enum

File: application/libraries/Sale_lib.php (hunk 14 above)
Comment (SteveIreland) [inline application/libraries/Sale_lib.php]: In a past life of developing software for international use I accustomed myself to using 0, 1, 2, etc. for state values instead of a character codes that insinuate a meaning that only make sense in a particular language.  However the negative side of that is that it obviously makes the code a little more difficult to read if one doesn't remember to add comments.

I can give the php-enum a shot.  Do I need to wait on the blessings of the Jedi Grand Master before starting to incorporate it?

**Comment by daN4cat** [comment]
php-enum incorporation needs composer run. I can do it in about 30min for you so it's all neat and clean.

**Comment by SteveIreland** [comment]
Sounds good to me.

**Comment by daN4cat** [comment]
@SteveIreland ok committed to master, but I used in the end this one: https://github.com/myclabs/php-enum.

It looks more popular.

You'll find a Code Igniter update, so no worries about the large rebase as it's all outside your code anyway so it should be straight forward.

Let me know how it goes with the enum class.

File: application/libraries/Tax_lib.php (hunk 12 above)
Comment (jekkos) [inline application/libraries/Tax_lib.php]: you could also just use define's, they map constant names to integers as well.

**Comment by jekkos** [comment]
@daN4cat @SteveIreland Could you check whether there's a possbility to use standard PHP language features like define instead of this custom enum thing? example as follows
```
define('CATEGORY', 0);
define('DROPDOWN', 1);
define('DATE', 2);
define('TEXT', 3);

const DEFINITION_TYPES = [CATEGORY, DROPDOWN, TEXT];
```

**Comment by SteveIreland** [comment]
Your name be on the package Captain @jekkos so I'm happy to do as you wish.

However, I think that the define statement makes the values global, so instead of referring to a class specific constant using `Rounding_code::HALF_UP` I would need to apply a naming convention such as `ROUNDING_CODE_HALF_UP` ... right?

There is a name space qualifier that can be applied, but my guess would be that it would be an inappropriate use of name spaces.

**Comment by jekkos** [comment]
Yes the addition of a namespace might make it even more useful. We could eg aggregate all constants in a file and then separate using namespaces. Also when denoting them it's directly clear to which domain it belongs.

**Comment by daN4cat** [comment]
Ok, it sounds we prefer define/constant instead of enumerations. It's fine that way too with me as long  we don't use magic numbers.

**Comment by daN4cat** [comment]
@SteveIreland I assume I can remove the enum package I added if you use the defines/constants.
I'll remove it over the week-end.

**Comment by SteveIreland** [comment]
You assume correctly.  Thanks.

**Comment by daN4cat** [comment]
@SteveIreland ok, removed from master

**Comment by SteveIreland** [comment]
I updated the pull request with numerous changes and also added new support for cash rounding.  Needs testing ... repeat ... needs testing.

I will, of course, continue to test these changes.  As you know, this impacts a lot.

I don't expect this to be the final modification to the pull request unless you deem it close enough to merge, in which case the turn around for any fixes will be quick.

**Comment by SteveIreland** [comment]
Uh .... ignore this latest update to the pull request.  While I was fixing issues for a VAT tax scenario or the existing sales tax scenario I inadvertently broke something on the new sales tax scenario.   Another one will be coming soon.

**Comment by odiea** [comment]
I just downloaded it and have the same db issues. Before I deleted my original database it would not recognize my current tax and the create tax modal would not drop down.

**Comment by odiea** [comment]
Sorry my bad I forgot you have to give permission for tax to show.
So db is ok

**Comment by SteveIreland** [comment]
@odiea Sorry about that.  To be clear, you're saying that when you click on "New Tax Code" that the Tax Code Form isn't appearing?  And does this occur on a new install or when you are upgrading an existing installation?

**Comment by odiea** [comment]
I am downloading your current code now. I had created a previous tax code and it did not show in the table view I then went to create a new tax and the modal just shows submit with no form.

**Comment by odiea** [comment]
The form problem is not present in this latest commit.

**Comment by SteveIreland** [comment]
Good!  Thanks.  Yes, that was the problem I discovered this morning I had to fix early this morning.  It took me forever to find the coding issue and figure out a resolution.

**Comment by odiea** [comment]
Taxes are not being totaled correctly in the register or on the receipt.

2017-04-02 18_29_08-open source point of sale _ powered by ospos 3 1 0
2017-04-02 18_30_14-open source point of sale _ powered by ospos 3 1 0

**Comment by SteveIreland** [comment]
So just to make sure I can replicate the settings for your test scenario.  You have "tax included" selected and the "customer sales tax" deselected.    What is your tax decimals, currency decimals, and cash decimals settings?

**Comment by SteveIreland** [comment]
Just for the record.  There are three tax scenarios that are supported.  1. Tax included, 2. Sales tax by item, and 3. Sales tax by customer.  I hoped to have multiple test iterations of each scenario  before submitting the pull request, but I came down with a nasty virus that knocked me for a loop and cost me the weekend.  So I'm playing catch up this week and appreciate the help.

**Comment by odiea** [comment]
I know its a pain but when little things like this show up I quit testing . You know what little things do.
I just created 1 sales tax and of course I ignored your wiki.
2017-04-03 17_11_44-open source point of sale _ powered by ospos 3 1 0
Here is what Takings shows.
2017-04-03 17_13_06-open source point of sale _ powered by ospos 3 1 0

**Comment by odiea** [comment]
I don't know if this is a fall back feature unless customer tax is enabled no tax is charged on a Sale. Should this still default to the default origin tax?

**Comment by SteveIreland** [comment]
Okay, thanks.  Allow me to run a few more debug cycles and will submit another pull request in a couple of days.

**Comment by SteveIreland** [comment]
The default origin tax is only used when the customer sales tax feature is enabled.

**Comment by odiea** [comment]
I get the same total issue with your feature disabled and include the tax in tax1 of item.
2017-04-03 17_49_36-open source point of sale _ powered by ospos 3 1 0

**Comment by SteveIreland** [comment]
I wonder if cash rounding is kicking in when you run your test.  Can you check the value of cash_decimals vs currency_decimals?

**Comment by odiea** [comment]
Are your future plan to eliminate the Item Tax% from showing on the Register when an Item has a tax set on it and customer tax is enabled. Currently I just see the tax amount removed.
2017-04-03 18_20_55-open source point of sale _ powered by ospos 3 1 0

**Comment by SteveIreland** [comment]
Correct.  The fact that the tax group name of "7.4% Sales Tax" is there is a bug which I shall fix.

**Comment by SteveIreland** [comment]
I forgot to initialize the cash_rounding config option to 2 so that by default cash_rounding is disabled.

```
INSERT INTO `ospos_app_config` (`key`, `value`) VALUES
  ('cash_decimals', '2');
```

**Comment by odiea** [comment]
Yes it was at 0 I changed it to2 dec and the tax is showing correctly as long as there is not tax set in item.
Another issue is now I have the tax set to 1 but it defaults to 2.
The takings is showing that it may be charging for both taxes when the item has a tax.
2017-04-03 18_51_56-open source point of sale _ powered by ospos 3 1 0
2017-04-03 18_53_05-open source point of sale _ powered by ospos 3 1 0

**Comment by odiea** [comment]
I am assuming now that the tax code should start at 0 and not 1.

**Comment by SteveIreland** [comment]
I'll check on the takings issue.  I wasn't expecting a problem there but it could because of the unwarranted tax group.

The default origin tax code is whatever code you assign to the tax code.  For example I could use MARY for the Maryland State tax code.  I would just type in MARY in the configuration.  I didn't add selection support for it, but if you think it should be added I can do so.

**Comment by odiea** [comment]
Ok I set the tax codes to names and still if there are only 2 codes created it still defaults to the 2nd tax code even if it was not the selected code. I created a 3rd tax code and the issue went away.

**Comment by SteveIreland** [comment]
The origin tax code (the one identified via configuration) is a "last resort".  The system will first try to identify the tax code to use based on the city and state of the customer using the city and/or state provided on the tax code configuration.  (Of course a tax code assigned to the customer overrides everything else).

Did you provide a city and state on the tax codes and do you have the same city and state that matches on the customer?  If you're only charging taxes on store sales and not delivering to the customer or providing services at a customer site then it only makes sense to set up an origin tax code.

**Comment by odiea** [comment]
Of course not. I am only setting different tax codes without selecting a customer to see what happens. I have found out that 60 to 70% of users do not read any manuals or wikis. As you can see by the issues.

**Comment by odiea** [comment]
This will be a great addon to OSPOS after people who want to use it actually set it up the way it is designed for. Great work and thanks for all your time.

**Comment by odiea** [comment]
Updating tax type in Item update does not change the Tax Category. I am now using an updated db.

**Comment by SteveIreland** [comment]
I just updated this feature so that I could start another test cycle.  My brain has been on hiatus due to a virus so I've lost track of what I've done over the last several days.  I'm just hoping I haven't done much damage.

File: application/libraries/Tax_lib.php (hunk 15 above)
Comment (daN4cat) [inline application/libraries/Tax_lib.php]: The code still has hardcoded values

File: application/libraries/Tax_lib.php (hunk 12 above)
Comment (daN4cat) [inline application/libraries/Tax_lib.php]: even here hardcoded values

File: application/controllers/Customers.php (hunk 4 above)
Comment (daN4cat) [inline application/controllers/Customers.php]: what about this?

**Comment by daN4cat** [comment]
@SteveIreland hope you are better. Whenever you are time please check the last three review comments, do some testing and then I'll merge this code.

Many thanks for the work done so far.

**Comment by odiea** [comment]
Still has issues with double taxation in All Reports and Takings.
Initial Sales looks ok when there is no Tax in the config. This is the first Receipt.
2017-04-08 06_52_34-wedidit _ powered by ospos 3 1 0
Here is the Takings.
2017-04-08 06_53_01-wedidit _ powered by ospos 3 1 0
Here is a Summary Report.

This is the Reprint Receipt.
2017-04-08 06_53_22-wedidit _ powered by ospos 3 1 0

**Comment by daN4cat** [comment]
Ok more to test and fix before we merge

**Comment by SteveIreland** [comment]
I don't think it is necessary to set pairs of constants for  yes/no, enabled/disabled, on/off, 1/0 etc since there are only two choices and they are all essentially Boolean.   However, I'm a geek and speak most derivatives of Boolean.  ;-)  So if you want to change all of the 0/1 tests in OSPOS I'd be happy to follow that standard going forward and making any corrections to existing code as time permits or working on a particular set of scripts.

**Comment by daN4cat** [comment]
For Boolean case you have TRUE/FALSE and the issue was more about 1 2 3 4 5....

**Comment by daN4cat** [comment]
The point about use !empty is to cover more cases with just one function call.

Anyway 0 / 1 is not really the point.

**Comment by SteveIreland** [comment]
@daN4cat Sorry ... my light-hearted response was to your question "What about this?" which had an example that was "Enabled/Disabled" using "1" or "0".

The previous question where you spotted my oversight on the $rounding_code has been completed (just not yet submitted).

I'm going to figure out what's going on with the doubling of the taxes that @odiea spotted before resubmitting the pull request. As you know (or can guess) it's  buried deep in the SQL and I just simply haven't been healthy enough to take it on until this week.

Stay tuned.

**Comment by daN4cat** [comment]
No worries, I often read github threads from my mobile so I have less context visibility.

Good job in general and thanks

**Comment by SteveIreland** [comment]
Corrections have been made to receipts, invoices, and other stuff.  Should be closer.

**Comment by SteveIreland** [comment]
I haven't yet had a chance to look at the double taxation, but when fixing a tax issue in takings I might have inadvertently fixed it.  I'll run a test on it tomorrow.

**Comment by odiea** [comment]
Sorry no such luck. If there is a tax value in the item I am still getting the double taxation in Takings and Reports.
2017-04-14 06_34_12-wedidit _ powered by ospos 3 1 0
In Takings the 1st sale has a tax in the Item and the second sale has the tax removed from item.
2017-04-14 06_35_48-wedidit _ powered by ospos 3 1 0

**Comment by SteveIreland** [comment]
I believe I've finally addressed the doubling up issue. It was a transition issue (one of many) where there are tax values for an item but the client is also enabling the new sales tax feature which isn't supposed to use the item tax entries.

File: application/models/Dinner_table.php [hunk 16]
```diff
@@ -57,17 +57,17 @@ public function get_empty_tables()
 
     public function get_name($dinner_table_id)
     {
-        $this->db->from('dinner_tables');
-        $this->db->where('dinner_table_id', $dinner_table_id);
-
-		$row = $this->db->get()->row();
-
-		if(isset($row->name))
+    	if($dinner_table_id == null || empty($dinner_table_id))
```
Comment (daN4cat) [inline application/models/Dinner_table.php]: null or empty is the same...

**Comment by daN4cat** [comment]
I merged this code.
I fixed few things here and there, mainly indentation ones, the empty from my comment and the database scripts as they were not aligned.
Please recheck my db changes to make sure they are correct.

**Comment by odiea** [comment]
@SteveIreland  You did a great job finding the double tax issue. It now coexists with or without customer tax enabled. In the Customer form is it suppose to retain the tax code if so that is not working.
